### PR TITLE
WIP: Revamp the cli, sdk

### PR DESCRIPTION
**Pull request by zephraph**
_Note_: The commit history for these changes got messed up due to an unfortunate rebase when trying to merge my changes into the new repo. Sorry about the lack of context there. 

This is pretty highly work in progress. 

## Getting started

To play around with this, you'll want to put both packages in watch mode. 

In one terminal

```
yarn workspace @linear/sdk dev
```

In another terminal

```
yarn workspace @linear/cli dev
```

Finally, from the root of the project run `yarn cli` to execute the cli. 

## SDK

The sdk exposes a thin graphql request wrapper for interacting with the api. It also exposes base primitives for interacting with things like issues, teams, etc. There's definitely more to be done here. 

One concept I'm playing with is the idea of a selection. When you're querying for a project/team/issue/etc there are generally different field that could result in the object you're searching for. For example, an issue will commonly be searched for by a key. I'll expand on this idea later in the PR. 

## CLI

This is _very_ bare bones at the moment and was just a mechanism to drive useful changes in the SDK. I'm happy with the base architecture of the CLI itself, but I'm not so sold on how the `create` command currently functions. 

One of the things that makes Linear stand apart is its user experience. I think the CLI should truly be an extension of that. Guided text interfaces could make it trivial to do things like change issue states or assign labels. Granted, a non-interactive mode is important to provide for automation too. More to come.

### CODE REVIEW

File: packages/sdk/README.md [hunk 1]
```diff
@@ -12,42 +12,10 @@ yarn install @linear/sdk
 import { Linear } from "@linear/sdk";
 
 const linear = new Linear({
-  token: "<your developer key>"
+  apiKey: "<your developer key>",
```
Comment (zephraph) [inline packages/sdk/README.md]: A small wording update to be more indicative to what the developer would see in the settings menu. Small things, but consistency really helps make tools approachable.

File: packages/sdk/src/gql.ts [hunk 2]
```diff
@@ -0,0 +1,3 @@
+import noop from "tagged-template-noop";
+
+export const gql = noop;
```
Comment (zephraph) [inline packages/sdk/src/gql.ts]: This is basically a helper to get editor highlighting.

File: packages/sdk/src/linear.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  public constructor({ apiKey, url = "https://api.linear.app/graphql" }: LinearArgs) {
+    this.client = new GraphQLClient(url, {
+      headers: {
+        Authorization: apiKey,
+        "Content-Type": "application/json",
+      },
+    });
+    this.issue = issue(this);
+    this.team = team(this);
+    this.label = label(this);
```
Comment (zephraph) [inline packages/sdk/src/linear.ts]: Heh, this is a bit of a... simplistic mixin system. I spent a little while faffing about with the types and trying to get something more elegant, but this isn't exactly the part of the project I wanted to spent a lot of time on so this works for now.

File: packages/sdk/src/models/issue.generated.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
+import * as Types from '../schema';
```
Comment (zephraph) [inline packages/sdk/src/models/issue.generated.ts]: All these `.generated.ts` files are generated automatically by `graphql-codegen`.

File: packages/sdk/src/models/issue.ts [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    ${IssueDetails}
+    query GetAllIssuesFromCLI {
+      issues {
+        ...IssueDetails
+      }
+    }
+  `);
+
+export const getIssue = async (client: Linear, selection: IssueSelection) => {
+  if ("id" in selection) {
```
Comment (zephraph) [inline packages/sdk/src/models/issue.ts]: This is an example of a selection in practice. An issue can be selected by `id` or `key`. This infrastructure wouldn't be necessary if and issue was queryable by both of these fields. 

Granted, GraphQL doesn't support input unions so there's no way in the schema to express that either `id` _or_ `key` is required. So very unfortunate...

Comment (jorilallo) [reply packages/sdk/src/models/issue.ts]: I think we could expand the API to support queries where one passes the key ID (XXX-123) as `id`, and same for teams as well. Then creating issues via the API would be much more simple

**Comment by jorilallo** [comment]
Lets land :)

**Comment by zephraph** [comment]
I'll wrap this up on Saturday! 👍

File: packages/sdk/src/models/issue.ts (hunk 5 above)
Comment (zephraph) [reply packages/sdk/src/models/issue.ts]: Yep!

**Comment by zephraph** [comment]
I'm doing a bit of a revamp. Ran into some issues w/ how oclif handles commands (can't have both a default command and subcommands, subcommands aren't flexible) so I changed the client and am focusing on getting the workflow.

Next steps:
- Flesh out any remaining v1 commands
- Wire up to sdk
- Build out test harness

Soon.

**Pull request merged**